client/src/components/EditorWithTabs/Editor/Monaco/languages/rust/rust-analyzer/rust-analyzer.ts

```diff
@@ -158,16 +160,20 @@ const loadLocalFiles = async () => {
  * @param model monaco editor model
  */
 const update = async (model: monaco.editor.IModel) => {
-  for (const crate of CRATES.importable) {
-    const status = cachedNames.get(crate);
-    if (status === "full") continue;
-
-    if (new RegExp(`${crate}::`, "gm").test(model.getValue())) {
-      await loadDependency(crate);
-    } else if (status !== "empty") {
-      await loadDependency(crate, { empty: true });
-    }
-  }
+  await Promise.all(
+    CRATES.importable.map((crate) => {
+      const status = cachedNames.get(crate);
+      if (status === "full") return null;
+
+      if (new RegExp(`${crate}::`, "gm").test(model.getValue())) {
+        return loadDependency(crate);
+      } else if (status !== "empty") {
+        return loadDependency(crate, { empty: true });
+      } else {
+        return null;
+      }
+    })
+  );
```

Comment:
For a simple benchmark, I've added `console.time("update");` [before here](https://github.com/solana-playground/solana-playground/blob/36bbb1a7e8db4760ef49a75d5a7cb5995192ce86/client/src/components/EditorWithTabs/Editor/Monaco/languages/rust/rust-analyzer/rust-analyzer.ts#L161) and `console.timeEnd("update");` [after here](https://github.com/solana-playground/solana-playground/blob/36bbb1a7e8db4760ef49a75d5a7cb5995192ce86/client/src/components/EditorWithTabs/Editor/Monaco/languages/rust/rust-analyzer/rust-analyzer.ts#L170).

**Previous:**

```
rust-analyzer.ts:172 update: 96.833984375 ms
rust-analyzer.ts:172 update: 0.086181640625 ms
rust-analyzer.ts:172 update: 0.093017578125 ms
rust-analyzer.ts:172 update: 0.406982421875 ms
rust-analyzer.ts:172 update: 0.31884765625 ms
rust-analyzer.ts:172 update: 0.06298828125 ms
rust-analyzer.ts:172 update: 0.067138671875 ms
rust-analyzer.ts:172 update: 0.35595703125 ms
rust-analyzer.ts:172 update: 0.092041015625 ms
rust-analyzer.ts:172 update: 0.06494140625 ms
rust-analyzer.ts:172 update: 0.130859375 ms
rust-analyzer.ts:172 update: 0.049072265625 ms
```


**After:**

```
rust-analyzer.ts:178 update: 49.904052734375 ms
rust-analyzer.ts:178 update: 0.131103515625 ms
rust-analyzer.ts:178 update: 5.697021484375 ms
rust-analyzer.ts:178 update: 19.4228515625 ms
rust-analyzer.ts:178 update: 7.8291015625 ms
rust-analyzer.ts:178 update: 15.10205078125 ms
rust-analyzer.ts:178 update: 12.961181640625 ms
rust-analyzer.ts:178 update: 6.27099609375 ms
rust-analyzer.ts:178 update: 2.1630859375 ms
rust-analyzer.ts:178 update: 7.06494140625 ms
rust-analyzer.ts:178 update: 7.393798828125 ms
rust-analyzer.ts:178 update: 7.012939453125 ms
```

While the first load is about 2x faster, each update is a lot slower compared to previous. `update` is called each time the content changes and I've tested with keystrokes. We could throttle the update but I didn't see a reason to do it because each update was already sub 1ms. I'm not sure why this is a lot slower but this is why optimimizing without benchmarking could be dangerous.

Please let me know if you have different results.

> **Note**
> The performance of [rust-analyzer](https://github.com/solana-playground/solana-playground/tree/d8f1696627ad7b70fe60c33c2c6c87c451296562/wasm/rust-analyzer) is not limited by these calls as the time it takes to load the crates is usually less than 1% of Rust Analyzer's initialization duration.

Reply:
Oh for sure. This is because you were able to synchronously escape the update if all the crates were `full`, and I made it so that you unconditionally wait until the next runloop.

Fixed in ea8c7eda0e597b86b106a0b12127cfa444bd5fcb.